### PR TITLE
feat: add broken-reference, unlinked-reference, and placeholder-text rules

### PR DESCRIPTION
**Pull request by not-stbenjam**
## Summary
- Add `content-broken-internal-reference` rule — detects markdown links pointing to nonexistent files (WARNING)
- Add `content-unlinked-internal-reference` rule — detects bare path-like strings not wrapped in links, configurable via `patterns` (INFO)
- Add `content-placeholder-text` rule — detects TODO markers, bracket placeholders, and unfilled templates (WARNING)

Based on ecosystem survey findings across 45 repositories (~2,100 skills).

## Test plan
- [ ] Tests for broken internal reference detection (existing files, missing files, URLs, template dirs)
- [ ] Tests for unlinked internal reference detection (bare paths, linked paths, custom patterns)
- [ ] Tests for placeholder text detection (TODO, FIXME, bracket placeholders)
- [ ] `make test` passes
- [ ] `make lint` passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added three content-quality checks: broken-internal-reference, unlinked-internal-reference (with configurable patterns), and placeholder-text; each has default severity and auto-enable settings.

* **Documentation**
  * Updated docs and generated docs tooling to describe the new rules, their default severities, autofix support, and unlinked-reference pattern defaults.

* **Tests**
  * Added comprehensive tests covering detection, exclusions, line reporting, configuration, and edge cases.

* **Chores**
  * Package and example config version bumped to 0.9.0; project scaffold now adjusts rule defaults (placeholder disabled by default).

[Review Change Stack](https://app.coderabbit.ai/change-stack/stbenjam/skillsaw/pull/153)

### CODE REVIEW

File: src/skillsaw/rules/builtin/content_rules.py [hunk 1]
```diff
@@ -1065,3 +1067,230 @@ def check(self, context: RepositoryContext) -> List[RuleViolation]:
                     violations.append(self.violation(msg, file_path=fpath))
 
         return violations
+
+
+class ContentBrokenInternalReferenceRule(Rule):
+    """Detect markdown links pointing to nonexistent files"""
+
+    formats = None
+    since = "0.8.3"
```
Comment (stbenjam) [inline src/skillsaw/rules/builtin/content_rules.py]: Please bump to 0.9.0 for these new rules

File: src/skillsaw/rules/builtin/content_rules.py [hunk 2]
```diff
@@ -1065,3 +1067,230 @@ def check(self, context: RepositoryContext) -> List[RuleViolation]:
...
+    repo_types = {
+        RepositoryType.AGENTSKILLS,
+        RepositoryType.SINGLE_PLUGIN,
+        RepositoryType.MARKETPLACE,
+    }
```
Comment (stbenjam) [inline src/skillsaw/rules/builtin/content_rules.py]: Why not APM? or not DOT_CLAUDE? What about UNKNOWN? Is there really a case where this shouldn't apply?

File: src/skillsaw/rules/builtin/content_rules.py [hunk 3]
```diff
@@ -1065,3 +1067,230 @@ def check(self, context: RepositoryContext) -> List[RuleViolation]:
...
+    _PLACEHOLDER_PATTERNS = [
+        (re.compile(r"\bTODO\b"), "TODO marker"),
+        (re.compile(r"\bFIXME\b"), "FIXME marker"),
+        (re.compile(r"\bXXX\b"), "XXX marker"),
+        (re.compile(r"\[link\s+here\]", re.IGNORECASE), "Placeholder link"),
+        (re.compile(r"\[Insert\s+[^\]]+\]", re.IGNORECASE), "Insert placeholder"),
+        (re.compile(r"\[If\s+[^\]]+\]", re.IGNORECASE), "Conditional placeholder"),
+        (
+            re.compile(
+                r"\*(?:TBD|to be added|details to be added|content to be added|will be added as you use)[^*]*\*",
```
Comment (stbenjam) [inline src/skillsaw/rules/builtin/content_rules.py]: `will be added as you use` seems prone to false positives

File: README.md [hunk 4]
```diff
@@ -517,6 +517,9 @@ Rules that go beyond structural validation to analyze the *quality* of instructi
 | `content-embedded-secrets` | Detect potential API keys, tokens, and passwords in instruction files | error (auto) | llm |
 | `content-banned-references` | Detect banned or deprecated model names, APIs, and custom patterns | warning (auto) | llm |
 | `content-inconsistent-terminology` | Detect inconsistent terminology across instruction files (e.g., mixing 'directory' and 'folder') | info (auto) | llm |
+| `content-broken-internal-reference` | Detect markdown links where the target file does not exist | warning (auto) | - |
+| `content-unlinked-internal-reference` | Detect bare path-like strings not wrapped in markdown link syntax | info (auto) | - |
+| `content-placeholder-text` | Detect TODO markers, bracket placeholders, and unfilled template text | warning (auto) | - |
```
Comment (stbenjam) [inline README.md]: The research doc needs to be updated with justification for these content rules (mostly just good software engineering practices, I think)

File: README.md [hunk 5]
```diff
@@ -537,6 +540,12 @@ Rules that go beyond structural validation to analyze the *quality* of instructi
 | `banned` | Additional banned patterns as list of {pattern, message} dicts | `[]` |
 | `skip-builtins` | Disable built-in deprecated model/API checks | `false` |
 
+**`content-unlinked-internal-reference` parameters:**
+
+| Parameter | Description | Default |
+|-----------|-------------|---------|
+| `patterns` | Glob patterns for path-like strings to flag when unlinked | `["./**/*.*", "references/**/*.md"]` |
```
Comment (stbenjam) [inline README.md:608]: I do not see tests and integration tests for this configuration

File: src/skillsaw/rules/builtin/content_rules.py (hunk 1 above)
Comment (not-stbenjam) [reply src/skillsaw/rules/builtin/content_rules.py]: Done -- bumped to 0.9.0 in 68d07c7.

File: src/skillsaw/rules/builtin/content_rules.py (hunk 2 above)
Comment (not-stbenjam) [reply src/skillsaw/rules/builtin/content_rules.py]: Good point. Changed all three rules to `repo_types = None` so they apply universally, including APM, DOT_CLAUDE, CODERABBIT, and UNKNOWN. These content quality checks (broken links, unlinked references, placeholder text) are format-agnostic and should fire everywhere.

File: src/skillsaw/rules/builtin/content_rules.py (hunk 3 above)
Comment (not-stbenjam) [reply src/skillsaw/rules/builtin/content_rules.py]: Agreed -- removed the `will be added as you use` pattern entirely. The remaining placeholder patterns (`*TBD*`, `*to be added*`, `*details to be added*`, `*content to be added*`) are more precisely scoped. Added a test confirming 'will be added as you use' is no longer flagged.

File: README.md (hunk 5 above)
Comment (not-stbenjam) [reply README.md:608]: Added tests for the `patterns` configuration on `content-unlinked-internal-reference`: `test_custom_patterns_config` verifies that restricting patterns filters violations correctly, and `test_empty_patterns_config_no_violations` verifies an empty patterns list produces zero violations.

File: README.md (hunk 4 above)
Comment (not-stbenjam) [reply README.md]: Acknowledged. These three rules are motivated by standard software engineering practices: broken links are dead references that mislead readers, unlinked paths lack navigability, and placeholder text signals incomplete documentation. Will update the research doc in a follow-up if needed, but these rules are straightforward quality checks that don't need upstream spec justification.

Comment (stbenjam) [reply README.md]: You must update the doc.

Comment (not-stbenjam) [reply README.md]: Done — added research justification sections for all three new rules to `docs/designs/content-rules-research.md` in commit 1154ddc:

- **content-broken-internal-reference**: grounded in standard link-checking practices (W3C, Google tech writing); broken links in instruction files remove dependencies from the agent's decision-making chain.
- **content-unlinked-internal-reference**: bare path references are un-verifiable and rot silently as the repo evolves; link syntax enables tooling and navigation.
- **content-placeholder-text**: same pattern as ESLint's `no-warning-comments` and SonarQube's TODO tracking applied to instruction files; LLMs can't distinguish placeholder text from real instructions.

Also updated the cross-cutting papers table to reference the new rules.

File: docs/designs/content-rules-research.md [hunk 6]
```diff
@@ -25,6 +25,9 @@ This document explains **why each rule exists**, and what research supports it.
 - [content-embedded-secrets](#content-embedded-secrets)
 - [content-banned-references](#content-banned-references)
 - [content-inconsistent-terminology](#content-inconsistent-terminology)
+- [content-broken-internal-reference](#content-broken-internal-reference)
+- [content-unlinked-internal-reference](#content-unlinked-internal-reference)
+- [content-placeholder-text](#content-placeholder-text)
```
Comment (stbenjam) [inline docs/designs/content-rules-research.md:30]: Are any of these potentially autofixable, either deterministically or with the llm ?

Comment (not-stbenjam) [reply docs/designs/content-rules-research.md:30]: Good question. Here's my assessment:

- **content-broken-internal-reference**: Not autofixable deterministically — the linter knows the link is broken but can't know the correct target (file may have been renamed, moved, or deleted). An LLM could suggest likely replacements by searching for similar filenames in the repo, but confidence would be low.

- **content-unlinked-internal-reference**: **Yes, deterministically autofixable.** The bare path `src/config.yaml` can be mechanically wrapped as `[src/config.yaml](src/config.yaml)`. This is a straightforward text transformation.

- **content-placeholder-text**: Not autofixable deterministically — removing a TODO or filling in a placeholder requires understanding what content should replace it. An LLM autofix could work here (e.g., replacing `*TBD*` with actual content inferred from context), but that's a heavier lift and more error-prone.

So `content-unlinked-internal-reference` is the clear candidate for deterministic autofix. The other two could be LLM-assisted but would need careful scoping. Want me to add autofix support for the unlinked reference rule in this PR, or track it as a follow-up?

Comment (stbenjam) [reply docs/designs/content-rules-research.md:30]: Yes we should fix unlinked internal reference if the file exists

For content-broken-internal-reference, perhaps you should look for likely matches in the repo that are similar and offer an alternatnve? (e.g. file `debug-cluster/SKLLI.md` not found, did you mean `debug-cluster/SKILL.md`?)

Comment (stbenjam) [reply docs/designs/content-rules-research.md:30]: Also one clarification, if a user puts custom content paths, that are .txt or .html or some other format that isn't markdown, we shouldn't try to fix it.  Make sure to have a test case for this

Comment (not-stbenjam) [reply docs/designs/content-rules-research.md:30]: Implemented both in 3d8ce8f:

- **`content-unlinked-internal-reference`**: SAFE autofix — wraps bare paths in `[path](path)` when the file exists in the repo. Only applies to verified-existing files, so confidence is SAFE.
- **`content-broken-internal-reference`**: SUGGEST autofix — uses exact filename matching first (catches file moves), then falls back to `difflib.get_close_matches` for fuzzy matching (catches typos). Violation messages now include 'did you mean ...' when a likely match is found. Confidence is SUGGEST since the match may not be the intended target.

Tests added for both. All 1181 tests pass.
